---
title: "Introduction to smoothLRC"
author: "Alexander White, Pengtao Dang, Changlin Wan, Anru Zhang, Chi Zhang, Sha Cao"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to smoothLRC}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Setup

smoothLRC is a novel tool for clustering spatial transcriptomics data. This technique utilizes a regularized maximum likelihood estimator to approximate the noisy observed expression matrix to incorporate spatial information and address expression dropout. This approximation under the Poisson distribution is both low-rank and spatially smooth. Furthermore, smoothLRC enables spatial clustering by modeling a low-dimensional representation of the count-based gene expression matrix. Thus, the approach encourages neighboring cells to belong to the same cluster via penalized matrix decomposition.

smoothLRC employs the [SummarizedExperiment](https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html) framework for spatial transcriptomics data. In this example, we will cluster a dorsolateral prefrontal cortex (DLPFC) collected and annotated by [Maynard et al. (2020)](https://www.biorxiv.org/content/10.1101/2020.02.28.969931v1).

```{r Setup, message=FALSE, warning=FALSE, collapse=TRUE}
library(smoothLRC)
library(ggplot2)
library(dplyr)
library(SummarizedExperiment)
set.seed(1)
```

## Load Data

Using the [spatialLIBD](http://spatial.libd.org/spatialLIBD/) package, we'll load sample 151671 from the DLPFC data. 

```{r Load Data, message=FALSE, warning=FALSE}
sample <- "151671"

sce <- spatialLIBD::fetch_data(type = "sce")
sce <- sce[, sce$sample_name == sample]
```

## Expression Plot

We'll find a highly variable gene (HVG) for this sample and explore its expression in the spatial context.

```{r Raw Expression, fig.height=5, fig.width=5.5, message=FALSE, warning=FALSE}
hvg <- rownames(sce)[which.max(apply(assay(sce), 1, var))]
sce %>% 
  spatial_plot(feature_name = hvg)
```


## Clustering

Next, we'll use smoothLRC to perform low-rank clustering on the data. We recommend setting the maximum number of iterations to a large value (e.g. 1,000). smoothLRC will add the clustering results to the colData and the model results to the metadata of the SummarizedExperiment object.

```{r smoothLRC, message=FALSE, warning=FALSE}
sce <- smooth_lrc(sce, lambda = 5, k = 30, n_clust = 5, maxiter = 5)
```


## Visualization

We'll visualize the results of both the model and a pre-computed model with a larger number of iterations and compare them with the manual annotations.

```{r Visualize, fig.height=5, fig.width=5.5, message=FALSE, warning=FALSE}
clust_1000 <- read.csv("151671.csv", colClasses = "factor")$smooth_cluster_1000
manual_labels <- colData(sce)$layer_guess_reordered

sce %>% 
  spatial_plot(type = "cluster") + 
  ggtitle("smoothLRC (5 iterations)") 

sce %>% 
  spatial_plot(type = "cluster", labels = clust_1000) +
  ggtitle("smoothLRC (1000 iterations)")

sce %>% 
  spatial_plot(type = "cluster", labels = manual_labels) +
  ggtitle("Manual Annotation")


```

Finally, we can reexamine our HVG using the smoothed results from the model.

```{r Smoothed Expression, fig.height=5, fig.width=5.5, message=FALSE, warning=FALSE}
sce %>% 
  spatial_plot(feature_name = hvg, type = "smooth")
```


Reproducibility

```{r Reproducibility, collapse=TRUE}
sessionInfo()
```

